kazuki-toolkit/src/main/kotlin/com/anaplan/engineering/kazuki/toolkit/ISO8601/Time.kt

```diff
@@ -0,0 +1,162 @@
+
+    override fun compareTo(other: TimeInZone) = normalisedTime.compareTo(other.normalisedTime)
+
+    @FunctionProvider(TimeInZoneFunctions::class)
+    val functions: TimeInZoneFunctions
+
+    class TimeInZoneFunctions(private val timeInZone: TimeInZone) {
+
+        val toNormalisedDurationSinceFirstTime: () -> Duration = function(
+            command = { timeInZone.functions.normaliseTimeInZone().time.functions.toDurationSinceFirstTime() },
+            post = { result -> result.functions.toTimeAfterFirstTime() == timeInZone.normalisedTime }
+        )
+
+        private val normaliseTimeInZonePlus: (Duration, Duration) -> NormalisedTime = function(
+            command = { utcTimeDuration, offsetDuration ->
+                if (offsetDuration <= utcTimeDuration) mk_NormalisedTime(
+                    utcTimeDuration.functions.subtractDuration(offsetDuration).functions.toTimeAfterFirstTime(),
+                    PlusOrMinus.None
+                ) else mk_NormalisedTime(
+                    utcTimeDuration.functions.addDuration(OneDayDuration).functions.subtractDuration(offsetDuration).functions.toTimeAfterFirstTime(),
+                    PlusOrMinus.Plus
+                )
+            }
+        )
+        private val normaliseTimeInZoneMinus: (Duration, Duration) -> NormalisedTime = function(
+            command = { utcTimeDuration, offsetDuration ->
+                val adjusted = utcTimeDuration.functions.addDuration(offsetDuration)
+                if (adjusted < OneDayDuration) mk_NormalisedTime(
+                    adjusted.functions.toTimeAfterFirstTime(),
+                    PlusOrMinus.None
+                ) else mk_NormalisedTime(
+                    adjusted.functions.subtractDuration(OneDayDuration).functions.toTimeAfterFirstTime(),
+                    PlusOrMinus.Minus
+                )
+            }
+        )
+
+        val normaliseTimeInZone: () -> NormalisedTime = function<NormalisedTime>(
+            command = {
+                val utcTimeDuration = timeInZone.time.functions.toDurationSinceFirstTime()
+                val offsetDuration = timeInZone.offset.offsetDuration
+                val directionOfOffset = timeInZone.offset.offsetDirection
+                when (directionOfOffset) {
+                    PlusOrMinus.Plus -> normaliseTimeInZonePlus(utcTimeDuration, offsetDuration)
+                    PlusOrMinus.Minus -> normaliseTimeInZoneMinus(utcTimeDuration, offsetDuration)
+                    PlusOrMinus.None -> mk_NormalisedTime(timeInZone.time, PlusOrMinus.None)
+                }
+            }
+        )
+        val format: () -> String = function<String>(
+            command = {
+                timeInZone.time.functions.format() +
+                        if (timeInZone.offset.offsetDuration != NoDuration) timeInZone.offset.functions.format() else "Z"
+            }
+        )
+    }
+}
+
+@Module
+interface NormalisedTime {
+    val time: Time
+    val plusOrMinusADay: PlusOrMinus
```

Comment:
I'm a bit confused about what `PlusOrMinus` is ..is plus or minus a day or a offset direction?

Reply:
PlusOrMinus is an enum class that can take one of three values: Plus, Minus or None
It is used both to indicate that midnight has been crossed when normalising a `TimeInZone` (`plusOrMinusADay`) and also the direction of an `Offset` (`offsetDirection`)

Reply:
Ambiguity resolved with a new typealias